File: src/android/com/ubudu/cordova/sdk/UbuduSDKCordova.java
//
// Copyright (c) 2011-2015, UBUDU SAS
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice, this
// list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//         SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

package com.ubudu.cordova.sdk;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import org.json.JSONException;
import org.json.JSONArray;
import org.json.JSONObject;

import android.util.Log;
import android.content.Context;

import com.ubudu.sdk.UbuduSDK;
import com.ubudu.sdk.UbuduBeaconManager;
import com.ubudu.sdk.UbuduGeofenceManager;

import com.ubudu.cordova.sdk.UbuduUserCordova;

public class UbuduSDKCordova extends CordovaPlugin {

    private boolean beaconsEnabled;
    private boolean geofencesEnabled;
    private boolean isRunning;

    @Override
    public void pluginInitialize() {
        this.beaconsEnabled = true;
        this.geofencesEnabled = true;
        this.isRunning = false;
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        Log.d("UbuduSDKCordova", "execute: action = " + action + " # args = " + args.length());
        
        if (action.equals("getSDKVersion")) {
            this.getSDKVersion(callbackContext);
        }
        else if (action.equals("getAppNamespace")) {
            this.getAppNamespace(callbackContext);
        }
        else if (action.equals("setAppNamespace")) {
            String appNamespace = args.getString(0);
            this.setAppNamespace(appNamespace, callbackContext);
        }
        else if (action.equals("getBaseURL")) {
            this.getBaseURL(callbackContext);
        }
        else if (action.equals("setBaseURL")) {
            String baseURL = args.getString(0);
            this.setBaseURL(baseURL, callbackContext);
        }
        else if (action.equals("setDelegate")) {
            // TODO get delegate param
            this.setDelegate(callbackContext);
        }
        else if (action.equals("getBeaconsEnabled")) {
            this.getBeaconsEnabled(callbackContext);
        }
        else if (action.equals("setBeaconsEnabled")) {
            boolean isEnabled = args.getBoolean(0);
            this.setBeaconsEnabled(isEnabled, callbackContext);
        }
        else if (action.equals("getGeofencesEnabled")) {
            this.getGeofencesEnabled(callbackContext);
        }
        else if (action.equals("setGeofencesEnabled")) {
            boolean isEnabled = args.getBoolean(0);
            this.setGeofencesEnabled(isEnabled, callbackContext);
        }
        else if (action.equals("getFileLogEnabled")) {
            this.getFileLogEnabled(callbackContext);
        }
        else if (action.equals("setFileLogEnabled")) {
            boolean isEnabled = args.getBoolean(0);
            this.setFileLogEnabled(isEnabled, callbackContext);
        }
        else if (action.equals("deviceSupportsGeofences")) {
            this.deviceSupportsGeofences(callbackContext);
        }
        else if (action.equals("deviceSupportsBeacons")) {
            this.deviceSupportsBeacons(callbackContext);
        }
        else if (action.equals("setUserInfo")) {
            JSONObject userInfo = args.getJSONObject(0);
            this.setUserInfo(userInfo, callbackContext);
        }
        else if (action.equals("start")) {
            this.start(callbackContext);
        }
        else if (action.equals("stop")) {
            this.stop(callbackContext);
        }
        else if (action.equals("isRunning")) {
            this.isRunning(callbackContext);
        }
        else if (action.equals("resetCounters")) {
            this.resetCounters(callbackContext);
        }
        else if (action.equals("removeAllData")) {
            this.removeAllData(callbackContext);
        }
        else if (action.equals("getDebugFileContent")) {
            this.getDebugFileContent(callbackContext);
        }
        else if (action.equals("clearDebugFile")) {
            this.clearDebugFile(callbackContext);
        } else {
            Log.e("UbuduSDKCordova", "Unknown action received (action = " + action + ")");
            return false;
        }
        return true;
    }

    private void getSDKVersion(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getSDKVersion");

        UbuduSDK sdk = getUbuduSDK();
        String sdkVersion = sdk.getVersion();

        callbackContext.success(sdkVersion);
    }

    private void getAppNamespace(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getAppNamespace");

        UbuduSDK sdk = getUbuduSDK();
        String appNamespace = sdk.namespace();

        callbackContext.success(appNamespace);
    }

    private void setAppNamespace(String appNamespace, CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setAppNamespace");

        UbuduSDK sdk = getUbuduSDK();
        sdk.setNamespace(appNamespace);

        callbackContext.success();
    }

    private void getBaseURL(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getBaseURL");

        UbuduSDK sdk = getUbuduSDK();
        // TODO
        callbackContext.error("Not implemented");
    }

    private void setBaseURL(String baseURL, CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setBaseURL");

        UbuduSDK sdk = getUbuduSDK();
        // TODO
        callbackContext.error("Not implemented");
    }

    private void setDelegate(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setDelegate");

        UbuduSDK sdk = getUbuduSDK();
        // TODO
        callbackContext.error("Not implemented");
    }

    private void getBeaconsEnabled(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getBeaconsEnabled");
        callbackContext.success(this.beaconsEnabled ? 1 : 0);
    }

    private void setBeaconsEnabled(boolean isEnabled, CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setBeaconsEnabled");

        UbuduSDK sdk = getUbuduSDK();
        Context context = this.getContext();
        UbuduBeaconManager beaconManager = sdk.getBeaconManager();
        this.beaconsEnabled = isEnabled;

        if (beaconManager != null) {
            if (beaconManager.isMonitoring() && isEnabled == false) {
                beaconManager.stop(context);
            } else if (!beaconManager.isMonitoring() && this.isRunning && isEnabled == true) {
                Error error = beaconManager.start(context);
                if (error != null) {
                    // TODO handle error
                }
            }
        }

        callbackContext.success();
    }

    private void getGeofencesEnabled(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getGeofencesEnabled");

        UbuduSDK sdk = getUbuduSDK();
        callbackContext.success(this.geofencesEnabled ? 1 : 0);
    }

    private void setGeofencesEnabled(boolean isEnabled, CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setGeofencesEnabled");

        UbuduSDK sdk = getUbuduSDK();
        Context context = this.getContext();
        UbuduGeofenceManager geofenceManager = sdk.getGeofenceManager();
        this.geofencesEnabled = isEnabled;

        if (geofenceManager != null) {
            if (geofenceManager.isMonitoring() && isEnabled == false) {
                geofenceManager.stop(context);
            } else if (!geofenceManager.isMonitoring() && this.isRunning && isEnabled == true) {
                Error error = geofenceManager.start(context);
                if (error != null) {
                    // TODO handle error
                }
            }
        }

        callbackContext.success();
    }

    private void getFileLogEnabled(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getFileLogEnabled");

        UbuduSDK sdk = getUbuduSDK();
        // TODO
        callbackContext.error("Not implemented");
    }

    private void setFileLogEnabled(boolean isEnabled, CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setFileLogEnabled");

        UbuduSDK sdk = getUbuduSDK();
        sdk.setFileLogEnabled(isEnabled);

        callbackContext.success();
    }

    private void deviceSupportsGeofences(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "deviceSupportsGeofences");

        UbuduSDK sdk = getUbuduSDK();
        UbuduGeofenceManager geofenceManager = sdk.getGeofenceManager();
        boolean geofencesSupported = (geofenceManager != null);
        
        callbackContext.success(geofencesSupported ? 1 : 0);
    }

    private void deviceSupportsBeacons(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "deviceSupportsBeacons");

        UbuduSDK sdk = getUbuduSDK();
        UbuduBeaconManager beaconManager = sdk.getBeaconManager();
        boolean beaconsSupported = (beaconManager != null);

        callbackContext.success(beaconsSupported ? 1 : 0);
    }

    private void setUserInfo(JSONObject userInfo, CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "setUserInfo");

        UbuduSDK sdk = getUbuduSDK();
        UbuduUserCordova user = new UbuduUserCordova(userInfo);
        sdk.setUserInformation(user);

        callbackContext.success();
    }

    private void start(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "start");

        Context context = this.getContext();
        UbuduSDK sdk = getUbuduSDK();
        UbuduGeofenceManager geofenceManager = sdk.getGeofenceManager();
        UbuduBeaconManager beaconManager = sdk.getBeaconManager();

        if (geofenceManager != null && this.geofencesEnabled && !geofenceManager.isMonitoring()) {
            Error error = geofenceManager.start(context);
            if (error != null) {
                Log.e("UbuduSDKCordova", "start: Error with geofenceManager.start");
                callbackContext.error(error.getLocalizedMessage());
                return;
            }
        }
        if (beaconManager != null && this.beaconsEnabled && !beaconManager.isMonitoring()) {
            Error error = beaconManager.start(context);
            if (error != null) {
                if (geofenceManager != null && geofenceManager.isMonitoring()) {
                    geofenceManager.stop(context);
                }

                Log.e("UbuduSDKCordova", "start: Error with beaconManager.start");
                callbackContext.error(error.getLocalizedMessage());
                return;
            }
        }
        this.isRunning = true;

        callbackContext.success();
    }

    private void stop(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "stop");

        Context context = this.getContext();
        UbuduSDK sdk = getUbuduSDK();
        UbuduGeofenceManager geofenceManager = sdk.getGeofenceManager();
        UbuduBeaconManager beaconManager = sdk.getBeaconManager();

        if (geofenceManager != null && geofenceManager.isMonitoring()) {
            geofenceManager.stop(context);
        }
        if (beaconManager != null && beaconManager.isMonitoring()) {
            beaconManager.stop(context);
        }
        this.isRunning = false;

        callbackContext.success();
    }

    private void isRunning(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "isRunning");
        callbackContext.success(this.isRunning ? 1 : 0);
    }

    private void resetCounters(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "resetCounters");

        UbuduSDK sdk = getUbuduSDK();
        // TODO
        callbackContext.error("Not implemented");
    }

    private void removeAllData(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "removeAllData");

        UbuduSDK sdk = getUbuduSDK();
        // TODO
        callbackContext.error("Not implemented");
    }

    private void getDebugFileContent(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "getDebugFileContent");

        UbuduSDK sdk = getUbuduSDK();
        String debugFileContent = sdk.debugFileContent();

        callbackContext.success(debugFileContent);
    }

    private void clearDebugFile(CallbackContext callbackContext) {
        Log.d("UbuduSDKCordova", "clearDebugFile");

        UbuduSDK sdk = getUbuduSDK();
        Context context = this.getContext();
        sdk.clearDebugFile(context);

        callbackContext.success();
    }

    private Context getContext() {
        return this.cordova.getActivity().getApplicationContext();
    }

    private UbuduSDK getUbuduSDK() {
        Context context = this.getContext();
        UbuduSDK sdk = UbuduSDK.getSharedInstance(context);
        return sdk;
    }
}

File: src/ios/UbuduSDKCordova.m
//
// Copyright (c) 2011-2015, UBUDU SAS
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice, this
// list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//         SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

#import <UbuduSDK/UbuduSDK.h>

#import "UbuduSDKCordova.h"

@implementation UbuduSDKCordova

- (void)getSDKVersion:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getSDKVersion");
    NSString *sdkVersion = [UbuduSDK sharedInstance].SDKVersion;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:sdkVersion];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getAppNamespace:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getAppNamespace");
    NSString *appNamespace = [UbuduSDK sharedInstance].appNamespace;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:appNamespace];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setAppNamespace:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;

    if ([command.arguments count] == 0) {
        NSLog(@"UbuduSDKCordova: setAppNamespace => error, no parameter");
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"App namespace parameter not provided"];
    }
    else {
        NSString *appNamespace = [command.arguments objectAtIndex:0];
        [UbuduSDK sharedInstance].appNamespace = appNamespace;
        NSLog(@"UbuduSDKCordova: setAppNamespace '%@'", appNamespace);
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getBaseURL:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getBaseURL");
    NSString *baseURL = [UbuduSDK sharedInstance].baseURL;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:baseURL];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setBaseURL:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;

    if ([command.arguments count] == 0) {
        NSLog(@"UbuduSDKCordova: setBaseURL => error, no parameter");
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Base URL parameter not provided"];
    }
    else {
        NSString *baseURL = [command.arguments objectAtIndex:0];
        [UbuduSDK sharedInstance].baseURL = baseURL;
        NSLog(@"UbuduSDKCordova: setBaseURL '%@'", baseURL);
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setDelegate:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: setDelegate");
    // TODO Implement
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Not implemented"];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getBeaconsEnabled:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getBeaconsEnabled");
    BOOL beaconsEnabled = [UbuduSDK sharedInstance].beaconsEnabled;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:beaconsEnabled];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setBeaconsEnabled:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;

    if ([command.arguments count] == 0) {
        NSLog(@"UbuduSDKCordova: setBeaconsEnabled => error, no parameter");
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"'Beacons enabled' boolean parameter not provided"];
    }
    else {
        NSNumber *beaconsEnabled = [command.arguments objectAtIndex:0];
        [UbuduSDK sharedInstance].beaconsEnabled = [beaconsEnabled boolValue];
        NSLog(@"UbuduSDKCordova: setBeaconsEnabled '%@'", beaconsEnabled);
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getGeofencesEnabled:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getGeofencesEnabled");
    BOOL geofencesEnabled = [UbuduSDK sharedInstance].geofencesEnabled;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:geofencesEnabled];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setGeofencesEnabled:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;

    if ([command.arguments count] == 0) {
        NSLog(@"UbuduSDKCordova: setGeofencesEnabled => error, no parameter");
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"'Geofences enabled' boolean parameter not provided"];
    }
    else {
        NSNumber *geofencesEnabled = [command.arguments objectAtIndex:0];
        [UbuduSDK sharedInstance].geofencesEnabled = [geofencesEnabled boolValue];
        NSLog(@"UbuduSDKCordova: setGeofencesEnabled '%@'", geofencesEnabled);
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getFileLogEnabled:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getFileLogEnabled");
    BOOL fileLogEnabled = [UbuduSDK sharedInstance].fileLogEnabled;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:fileLogEnabled];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setFileLogEnabled:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;

    if ([command.arguments count] == 0) {
        NSLog(@"UbuduSDKCordova: setFileLogEnabled => error, no parameter");
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"'File log enabled' boolean parameter not provided"];
    }
    else {
        NSNumber *fileLogEnabled = [command.arguments objectAtIndex:0];
        [UbuduSDK sharedInstance].fileLogEnabled = [fileLogEnabled boolValue];
        NSLog(@"UbuduSDKCordova: setFileLogEnabled '%@'", fileLogEnabled);
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)deviceSupportsGeofences:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: deviceSupportsGeofences");
    CDVPluginResult* pluginResult = nil;

    NSError *error = nil;
    BOOL geofencesSupported = [UbuduSDK deviceSupportsGeofences:[UIDevice currentDevice] error:&error];

    if (error != nil) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:geofencesSupported];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)deviceSupportsBeacons:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: deviceSupportsBeacons");
    CDVPluginResult* pluginResult = nil;

    NSError *error = nil;
    BOOL beaconsSupported = [UbuduSDK deviceSupportsBeacons:[UIDevice currentDevice] error:&error];

    if (error != nil) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:beaconsSupported];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)start:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: start");
    CDVPluginResult* pluginResult = nil;
    
    NSError *error = nil;
    BOOL result = [[UbuduSDK sharedInstance] start:&error];
    if (result) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        if (error != nil) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        }
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)stop:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: stop");
    [[UbuduSDK sharedInstance] stop];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)isRunning:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: isRunning");
    BOOL isRunning = [UbuduSDK sharedInstance].isRunning;

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:isRunning];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)resetCounters:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: resetCounters");
    CDVPluginResult* pluginResult = nil;
    
    NSError *error = nil;
    BOOL result = [[UbuduSDK sharedInstance] resetCounters:&error];
    if (result) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        if (error != nil) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        }
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)removeAllData:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: removeAllData");
    CDVPluginResult* pluginResult = nil;
    
    NSError *error = nil;
    BOOL result = [[UbuduSDK sharedInstance] removeAllData:&error];
    if (result) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        if (error != nil) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        }
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getDebugFileContent:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: getDebugFileContent");

    NSData *debugFileContent = [[UbuduSDK sharedInstance] getDebugFileContent];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArrayBuffer:debugFileContent];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)clearDebugFile:(CDVInvokedUrlCommand*)command
{
    NSLog(@"UbuduSDKCordova: clearDebugFile");

    [[UbuduSDK sharedInstance] clearDebugFile];
    [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK] callbackId:command.callbackId];
}

@end